### PR TITLE
Revert "Revert "Use brave search as a default search provider in tor window and update TorWindow NTP""

### PR DESCRIPTION
**Pull request by simonhong**
Reverts brave/brave-core#12390

First commit is reverting previsouly reverted one.
Second commit fixed below issues that has reverted code.
* Fixed default search engines list has two brave entries by hiding from settings UI.
* Fixed onion url is used for searching from normal window by using different keyword.
## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Launch browser and open tor window
2. Search anything from tor window location bar
3. Check onion url is used for searching
4. Go to normal window and set brave as a default search provider
5. Search anything from normal window location bar
6. Check search.brave.com is used for it
7. Re-launch and load brave://settings/searchEngines
8. Check default list has only one brave entry(search.brave.com)
9. Launch tor window and pase`:search.brave4u7jddbv7cyviptqjc7jusxh72uik7zt6adtckl5f4nwy2v72qd.onion` and press tab.
10. Check `:search.brave4u7jddbv7cyviptqjc7jusxh72uik7zt6adtckl5f4nwy2v72qd.onion` works as a provider keyword

### CODE REVIEW

**Comment by simonhong** [comment]
Not ready to review.

**Comment by simonhong** [comment]
Ready to review :)

**Review by emerick — APPROVED** [review]

File: browser/ui/webui/settings/brave_search_engines_handler.cc [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/ui/webui/settings/brave_search_engines_handler.cc]: Yup, updated :)

**Comment by simonhong** [comment]
Kindly ping to @brave/chromium-src-reviewers :)

File: browser/ui/webui/settings/brave_search_engines_handler.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#define BRAVE_BROWSER_UI_WEBUI_SETTINGS_BRAVE_SEARCH_ENGINES_HANDLER_H_
+
+#include <memory>
+
+#include "base/values.h"
+#include "chrome/browser/ui/webui/settings/search_engines_handler.h"
+
+namespace settings {
+
+class BraveSearchEnginesHandler : public settings::SearchEnginesHandler {
```
Comment (goodov) [inline browser/ui/webui/settings/brave_search_engines_handler.h:16]: just `public SearchEnginesHandler`

File: chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_WEBUI_SETTINGS_SEARCH_ENGINES_HANDLER_H_
+#define BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_WEBUI_SETTINGS_SEARCH_ENGINES_HANDLER_H_
+
+#include <memory>
+
+#include "base/values.h"
```
Comment (goodov) [inline chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h:11]: I don't think we need these includes.

File: chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+#define GetSearchEnginesList                          \
+  GetSearchEnginesList_Empty() {                      \
```
Comment (goodov) [inline chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h:14]: Usually `NotUsed` or something like that. Right now it looks like it's called in the base class.

File: chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    return std::make_unique<base::DictionaryValue>(); \
```
Comment (goodov) [inline chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h:15]: `nullptr` maybe?

File: browser/ui/webui/settings/brave_search_engines_handler.h (hunk 2 above)
Comment (simonhong) [reply browser/ui/webui/settings/brave_search_engines_handler.h]: Done.

File: chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h (hunk 3 above)
Comment (simonhong) [reply chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h]: Removed.

File: chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h (hunk 4 above)
Comment (simonhong) [reply chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h]: `NotUsed` seems much better.

File: chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h (hunk 5 above)
Comment (simonhong) [reply chromium_src/chrome/browser/ui/webui/settings/search_engines_handler.h]: Done.

File: browser/ui/webui/settings/brave_search_engines_handler.cc [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+std::unique_ptr<base::DictionaryValue>
+BraveSearchEnginesHandler::GetSearchEnginesList() {
+  auto search_engines_info = SearchEnginesHandler::GetSearchEnginesList();
+  // Don't show two brave search entries from settings to prevent confusion.
+  // Hide brave search for tor entry from settings UI. User doesn't need to
+  // select brave search tor entry for normal profile.
+  constexpr char kDefaultsKey[] = "defaults";
+  auto* defaults = search_engines_info->FindListKey(kDefaultsKey);
+  DCHECK(defaults && defaults->is_list());
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/ui/webui/settings/brave_search_engines_handler.cc]: removed.

**Review by goodov — APPROVED** [review]

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 11 x64` using the following build(s):

```
Brave | 1.38.39 Chromium: 99.0.4844.51 (Official Build) nightly (64-bit)
-- | --
Revision | d537ec02474b5afe23684e7963d538896c63ac77-refs/branch-heads/4844@{#875}
OS | Windows 11 Version 21H2 (Build 22000.556)
```

Verified the STR/Cases mentioned via https://github.com/brave/brave-core/pull/12391#issue-1148744623. Also went through the cases mentioned below.

#### **`Test Case #1`** - Clean Install

* ensured that Brave Search is the default SE for both Normal/Private windows
* ensured that Brave Search is the default SE for Tor Windows and ensured that the `.onion` address is being used
  * `https://search.brave4u7jddbv7cyviptqjc7jusxh72uik7zt6adtckl5f4nwy2v72qd.onion/search?q`
* ensured that `https://search.brave.com/search?q=` is still being used under Normal & Private windows
* ensured that there's no mention of DDG anywhere under the Tor window nor any images associated with DDG
* ensured that the progress bar correctly indicates when Tor is being downloaded/connecting and connected successfully
* ensured that Brave Search is only listed once under `brave://settings/search` & `brave://settings/searchEngines`
* ensured that the `:search.brave4u7jddbv7cyviptqjc7jusxh72uik7zt6adtckl5f4nwy2v72qd.onion` keyword is working as expected
* ensured that changing defaults via brave://settings/searchEngines doesn't affect the default Tor SE (should always be Brave)

**`Example`** | **`Example`** | **`Example`** | **`Example`** | **`Example`**
----------------|---------------|-----------------|----------------|-----------------
torDownloading | torSuccess | defaultSearch |keyword1 | image

**`NOTE:`** you'll most likely run into several CAPTCHA screens whenever searching via the Tor window. Double checking via https://bravesoftware.slack.com/archives/C01NNFM9XMM/p1647190731570629 but I don't think there's anything we can do as this is pretty much expected whenever using a Tor window.

CAPTCHA

#### **`Test Case #2`** - Upgrade (Other SE default other than Brave Search)

* downloaded/installed `1.38.28 Chromium: 99.0.4844.51`
* ensured that the Tor window mentions that DDG is the default SE
* ensured that the Tor window is using `https://duckduckgo.com/?q=` as the default SE
* changed the default SE to `Startpage` via `brave://settings/search`
* ensured that `Startpage` is set as the default SE via `brave://settings/searchEngines`
* Upgraded to `1.38.39 Chromium: 99.0.4844.51`
* ensured that `Startpage` is still the default SE under `brave://settings/search` & `brave://settings/searchEngines`
* ensured that there's no duplicates being listed under `brave://settings/search` & `brave://settings/searchEngines`
* ensured that all mentions of DDG are removed from the Tor window
* ensured that `https://search.brave4u7jddbv7cyviptqjc7jusxh72uik7zt6adtckl5f4nwy2v72qd.onion/search?q=` is the default SE

**`Before`** - `1.38.28 Chromium: 99.0.4844.51`

**`Example`** | **`Example`** | **`Example`**
----------------|----------------|----------------
beforeTor | beforeSE1 | beforeSE2

**`After`** - `1.38.39 Chromium: 99.0.4844.51`

**`Example`** | **`Example`** | **`Example`** | **`Example`** | **`Example`**
----------------|----------------|----------------|----------------|-----------------
image | image | image | image | image

#### **`Test Case #3`** - Upgrade (Brave Search as default)

**Pull request merged**